File: 백준/Silver/15666. N과 M （12）/N과 M （12）.cc
#include<iostream>
#include<vector>
#include<algorithm>
using namespace std;

int N, M;
vector<int> arr;
vector<int> input;

void dfs(int n, int k) {
    if (k == M) {
        for (auto i = 0; i < M; ++i) {
            cout << arr[i] << " ";
        }
        cout << "\n";

        return;
    }
    int num = -1;
    for (auto i = n; i <= N; ++i) {
        if (num != input[i]) {
            arr.push_back(input[i]);
            num = input[i];
            dfs(i,k + 1);
            arr.pop_back();
        }
    }
}

int main(void) {
    cin.tie(0)->sync_with_stdio(0);

    cin >> N >> M;

    input.push_back(0);

    for (auto i = 0; i < N; ++i) {
        int a;
        cin >> a;
        input.push_back(a);
    }

    sort(input.begin(), input.end());
    dfs(1,0);

    return 0;
}

File: 백준/Silver/15666. N과 M （12）/N과 M （12）.java
import java.util.*;
import java.io.*;


public class Main {
    static int[] input, arr;
    static int N, M;
    static boolean[] vis;
    static StringBuilder sb = new StringBuilder();

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());

        arr = new int[M];
        vis = new boolean[N];
        input = new int[N];
        st = new StringTokenizer(br.readLine());

        for (int i = 0; i < N; i++) {
            input[i] = Integer.parseInt(st.nextToken());
        }
        Arrays.sort(input);

        DFS(0, 0);

        System.out.println(sb.toString());
    }

    static void DFS(int depth, int cur) {
        if (depth == M) {
            for (int n : arr) {
                sb.append(n).append(" ");
            }
            sb.append("\n");
            return;
        }

        int temp = -1;
        for (int i = cur; i < N; i++) {
            int now = input[i];
            if (temp == now) continue;
            else {
                temp = now;
                arr[depth] = now;
                DFS(depth + 1, i);
            }
        }
    }
}
